src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -2712,7 +2726,17 @@ public void visitForLoop(JCForLoop tree) {
                     uninitsSkip.assign(uninits);
                     uninitsSkip.inclRange(firstadr, nextadr);
                 }
-                scan(tree.body);
+                final Bits reassignedInForLoopBodyPrev = reassignedInForLoopBody;
```

Comment:
My memory of this code is a bit rusty, so please bear with me... I seem to recall we mark all variables as effectively final, and then "unmark" them if see see them in `letInit`. If this is correct, could we simply re-force the effectively final flag of a loop variable before the loop body?

Reply:
> could we simply re-force the effectively final flag of a loop variable before the loop body?

That's perfectly natural idea and at first I even tried it :)

The problem is that `AssignAnalyzer` is where DA/DU analysis happens, "effectively final" is computed, and where `letInit()` lives, but `CaptureAnalysis` is where we actually test for whether a lambda may capture a variable, and those are currently two independent steps, so it can't all be done "inline" in one pass.

Since "effectively final" is a global property of a variable that is independent of location, whereas the property we want to check is more like "effectively final in the associated loop body" (aka. "not reassigned in the loop body"), we need a new flag in order to pass along this new information.

Side note: flags seem in short supply, and both of these flags are only used within `Flow.java`, so instead of using flags we could keep this information private to `Flow.java`, e.g.:
```java
    AssignAnalyzer aa = new AssignAnalyzer();
    aa.analyzeTree(env, make);
    ...
    new CaptureAnalyzer(aa).analyzeTree(env, make);
    ...
```
This would better follow the good housekeeping principle of limiting state to the minimum possible scope.

Reply:
> That's perfectly natural idea and at first I even tried it :)
> 
> The problem is that `AssignAnalyzer` is where DA/DU analysis happens, "effectively final" is computed, and where `letInit()` lives, but `CaptureAnalysis` is where we actually test for whether a lambda may capture a variable, and those are currently two independent steps, so it can't all be done "inline" in one pass.

I see the problem - thanks for the explanation. Now that I remember more of the code, there's a reason why effective finality is determined as part of DA/DU - that has to do with some enhancements we did in Java 8 to allow for "blank" effectively final variables. That is a variable like this:

```
int i;
if (cond) {
   i = 2;
} else {
   i = 3;
}
```

Is considered effectively final (unless is further mutated), even though it lacks an initializer and there is some control flow. So we settled to do that in the two passes you describe.